source/common/filesystem/posix/directory_iterator_impl.cc

```diff
@@ -42,39 +42,39 @@ void DirectoryIteratorImpl::nextEntry() {
   }
 
   if (entry == nullptr) {
-    entry_ = {"", FileType::Other};
+    entry_ = {"", FileType::Other, 0};
   } else {
     const std::string current_path(entry->d_name);
     const std::string full_path(directory_path_ + "/" + current_path);
-    entry_ = {current_path, fileType(full_path, os_sys_calls_)};
+    entry_ = makeEntry(entry->d_name);
   }
 }
 
-FileType DirectoryIteratorImpl::fileType(const std::string& full_path,
-                                         Api::OsSysCallsImpl& os_sys_calls) {
+DirectoryEntry DirectoryIteratorImpl::makeEntry(absl::string_view filename) const {
+  const std::string full_path = absl::StrCat(directory_path_, "/", filename);
   struct stat stat_buf;
+  FileType file_type = FileType::Other;
 
-  const Api::SysCallIntResult result = os_sys_calls.stat(full_path.c_str(), &stat_buf);
+  const Api::SysCallIntResult result = os_sys_calls_.stat(full_path.c_str(), &stat_buf);
   if (result.return_value_ != 0) {
-    if (errno == ENOENT) {
+    if (result.errno_ == ENOENT) {
       // Special case. This directory entity is likely to be a symlink,
       // but the reference is broken as the target could not be stat()'ed.
       // If we confirm this with an lstat, treat this file entity as
       // a regular file, which may be unlink()'ed.
       if (::lstat(full_path.c_str(), &stat_buf) == 0 && S_ISLNK(stat_buf.st_mode)) {
-        return FileType::Regular;
+        file_type = FileType::Regular;
       }
     }
-    throw EnvoyException(fmt::format("unable to stat file: '{}' ({})", full_path, errno));
-  }
-
-  if (S_ISDIR(stat_buf.st_mode)) {
-    return FileType::Directory;
+    if (file_type == FileType::Other) {
+      throw EnvoyException(fmt::format("unable to stat file: '{}' ({})", full_path, errno));
```

Comment:
pre-existing, but can makeEntry be called during a request? It seems like it can, in which case we should switch from throwing an exception to returning an absl::StatusOr. It is not permitted to throw exceptions on worker threads in Envoy.

Actually throwing them is OK if they crash the program (for some definition of OK). But we disallow *catch* in worker threads.

This is beyond the scope of this PR but would it be worth it to throw in a TODO while you are here?

Reply:
Added TODO. (I too thought throwing an exception was a weird choice here.)